In [ ]:
pip install aif360 BlackBoxAuditing

Looking in indexes: https://pypi.org/simple, https://us-python.pkg.dev/colab-wheels/public/simple/
     ━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━ 214.1/214.1 kB 5.2 MB/s eta 0:00:00
     ━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━ 2.6/2.6 MB 40.3 MB/s eta 0:00:00
  Preparing metadata (setup.py) ... done
  Created wheel for BlackBoxAuditing: filename=BlackBoxAuditing-0.1.54-py2.py3-none-any.whl size=1394768 sha256=78b8f76908d8d12d16d20ea003d0b5c3aa5ab64b233328b4040ade263fb76fd7
  Stored in directory: /root/.cache/pip/wheels/8f/3c/f8/2ad8792a15548dfb008ec5738566ea9e5aa8999311732473fa
Successfully built BlackBoxAuditing


In [ ]:
import pandas as pd
import aif360
import numpy as np
from aif360.datasets import BinaryLabelDataset
from aif360.algorithms.preprocessing import DisparateImpactRemover
from sklearn.model_selection import train_test_split
from sklearn.linear_model import LogisticRegression
from random import randrange

pip install 'aif360[LawSchoolGPA]'
pip install 'aif360[Reductions]'
pip install 'aif360[Reductions]'
pip install 'aif360[Reductions]'


In [ ]:
csv_path = "StudentPerformanceWithNewCoumns.csv"

In [ ]:
data = pd.read_csv(csv_path).drop(columns=['Unnamed: 0'])
data.head()

gender     race parental_level_of_education         lunch  \
0  female  group B           bachelor's degree      standard   
1  female  group C                some college      standard   
2  female  group B             master's degree      standard   
3    male  group A          associate's degree  free/reduced   
4    male  group C                some college      standard   

  test_preparation_course  math_score  reading_score  writing_score  \
0                    none          72             72             74   
1               completed          69             90             88   
2                    none          90             95             93   
3                    none          47             57             44   
4                    none          76             78             75   

  overall_result math_result reading_result writing_result  
0           Pass        Pass           Pass           Pass  
1           Pass        Pass           Pass           Pass  
2           Pass        Pass           Pass           Pass  
3           Fail        Fail           Fail           Fail  
4           Pass        Pass           Pass           Pass

# STEP 3

## 1) Based on your dataset, identify the privileged/unprivileged groups associated with each of your protected class variables

In [ ]:

## protected class variables are:
# gender
# race

## 2) For each protected class variable, select two fairness metrics and compute the fairness metrics associated with your privileged/unprivileged groups as a function of each of your two dependent variables. You may choose any reasonable threshold in order to generate a baseline for comparison using the fairness metrics.## 

In [ ]:
# fairness metrics used - Disparate Impact Ratio and Statistical Parity Difference  
# math_result and reading_result used as the 2 dependent variables (threshold of 60 / 100)

#### Gender - Math result

In [ ]:
# for dependent variable of math_result
print(data.groupby(["gender", "math_result"]).agg({"parental_level_of_education": "count"}))
print()
# Disparate Impact Ratio

favarouble_unpriv = 321 / (197 + 321)
favarouble_priv = 356 / (356 + 126)

_dir = favarouble_unpriv / favarouble_priv
print("DIR: ", _dir)

# Statistical Parity Difference

_spd = (favarouble_unpriv * 100) - (favarouble_priv * 100)
print("SPD: ", _spd)

                    parental_level_of_education
gender math_result                             
female Fail                                 197
       Pass                                 321
male   Fail                                 126
       Pass                                 356

DIR:  0.8390199991323586
SPD:  -11.88980919271377


#### Gender - reading result

In [ ]:
# for dependent variable of reading_result
print(data.groupby(["gender", "reading_result"]).agg({"parental_level_of_education": "count"}))
print()
# Disparate Impact Ratio

favarouble_unpriv = 427 / (91 + 427)
favarouble_priv = 319 / (319 + 163)

_dir = favarouble_unpriv / favarouble_priv
print("DIR: ", _dir)

# Statistical Parity Difference

_spd = (favarouble_unpriv * 100) - (favarouble_priv * 100)
print("SPD: ", _spd)

                       parental_level_of_education
gender reading_result                             
female Fail                                     91
       Pass                                    427
male   Fail                                    163
       Pass                                    319

DIR:  1.2455307972549352
SPD:  16.249859818324552


#### Gender - Overall_result

In [ ]:
# for dependent variable of overall_result
print(data.groupby(["gender", "overall_result"]).agg({"parental_level_of_education": "count"}))
print()
# Disparate Impact Ratio

favarouble_unpriv = 425 / (93 + 425)
favarouble_priv = 301 / (301 + 181)

_dir = favarouble_unpriv / favarouble_priv
print("DIR: ", _dir)

# Statistical Parity Difference

_spd = (favarouble_unpriv * 100) - (favarouble_priv * 100)
print("SPD: ", _spd)

                       parental_level_of_education
gender overall_result                             
female Fail                                     93
       Pass                                    425
male   Fail                                    181
       Pass                                    301

DIR:  1.3138316294462473
SPD:  19.59819926624907


#### Race - math result

In [ ]:
# there are 5 groups of race (A to E)
# i'm calling group C as unprivileged and others are priviledged

data.loc[data.race == 'group A', 'race'] = 1
data.loc[data.race == 'group B', 'race'] = 1
data.loc[data.race == 'group C', 'race'] = 0
data.loc[data.race == 'group D', 'race'] = 1
data.loc[data.race == 'group E', 'race'] = 1

In [ ]:
# for dependent variable of math_result
print(data.groupby(["race", "math_result"]).agg({"parental_level_of_education": "count"}))
print()
# Disparate Impact Ratio

favarouble_unpriv = 206 / (206 + 113)
favarouble_priv = 471 / (471 + 210)

_dir = favarouble_unpriv / favarouble_priv
print("DIR: ", _dir)

# Statistical Parity Difference

_spd = (favarouble_unpriv * 100) - (favarouble_priv * 100)
print("SPD: ", _spd)

                  parental_level_of_education
race math_result                             
0    Fail                                 113
     Pass                                 206
1    Fail                                 210
     Pass                                 471

DIR:  0.9336900744763692
SPD:  -4.586193086876676


#### race - reading result

In [ ]:
# for dependent variable of reading_result
print(data.groupby(["race", "reading_result"]).agg({"parental_level_of_education": "count"}))
print()
# Disparate Impact Ratio

favarouble_unpriv = 244 / (244 + 75)
favarouble_priv = 502 / (502 + 179)

_dir = favarouble_unpriv / favarouble_priv
print("DIR: ", _dir)

# Statistical Parity Difference

_spd = (favarouble_unpriv * 100) - (favarouble_priv * 100)
print("SPD: ", _spd)

                     parental_level_of_education
race reading_result                             
0    Fail                                     75
     Pass                                    244
1    Fail                                    179
     Pass                                    502

DIR:  1.037630044086975
SPD:  2.7739033967197315


## 3) Select a pre-processing bias mitigation algorithm to transform the original dataset (e.g. Reweighting, Disparate Impact Remover, etc.) as a function of one of your dependent variables

In [ ]:
# bias mitigation algorithm - DisparateImpactRemover
# dependent variables - math_result

In [ ]:
data.loc[data.gender == 'male', 'gender'] = 1
data.loc[data.gender == 'female', 'gender'] = 0

# data.loc[data.reading_result == 'Pass', 'reading_result'] = 1
# data.loc[data.reading_result == 'Fail', 'reading_result'] = 0

# data.loc[data.writing_result == 'Pass', 'writing_result'] = 1
# data.loc[data.writing_result == 'Fail', 'writing_result'] = 0

data.loc[data.math_result == 'Pass', 'math_result'] = 1
data.loc[data.math_result == 'Fail', 'math_result'] = 0

data.loc[data.reading_result == 'Pass', 'reading_result'] = 1
data.loc[data.reading_result == 'Fail', 'reading_result'] = 0

data.loc[data.writing_result == 'Pass', 'writing_result'] = 1
data.loc[data.writing_result == 'Fail', 'writing_result'] = 0

data.loc[data.overall_result == 'Pass', 'overall_result'] = 1
data.loc[data.overall_result == 'Fail', 'overall_result'] = 0

data.loc[data.test_preparation_course == 'completed', 'test_preparation_course'] = 1
data.loc[data.test_preparation_course == 'none', 'test_preparation_course'] = 0

# data.loc[data.overall_result == 'Pass', 'overall_result'] = 1
# data.loc[data.overall_result == 'Fail', 'overall_result'] = 0

In [ ]:
data.head()

gender race parental_level_of_education         lunch  \
0      0    1           bachelor's degree      standard   
1      0    0                some college      standard   
2      0    1             master's degree      standard   
3      1    1          associate's degree  free/reduced   
4      1    0                some college      standard   

  test_preparation_course  math_score  reading_score  writing_score  \
0                       0          72             72             74   
1                       1          69             90             88   
2                       0          90             95             93   
3                       0          47             57             44   
4                       0          76             78             75   

  overall_result math_result reading_result writing_result  
0              1           1              1              1  
1              1           1              1              1  
2              1           1              1              1  
3              0           0              0              0  
4              1           1              1              1

In [ ]:

binaryLabelDataset = BinaryLabelDataset(
    favorable_label=1,
    unfavorable_label=0,
    df=data[['gender','race','overall_result','test_preparation_course','math_score','reading_score','writing_score','math_result','reading_result','writing_result']],
    label_names=['math_result'],
    protected_attribute_names=['gender'],
)

In [ ]:
di = DisparateImpactRemover(repair_level=1, sensitive_attribute='gender')
transformed_data = di.fit_transform(binaryLabelDataset)
transformed = transformed_data.convert_to_dataframe()[0]
transformed

gender  race  overall_result  test_preparation_course  math_score  \
0       0.0   1.0             1.0                      0.0        72.0   
1       0.0   0.0             1.0                      1.0        68.0   
2       0.0   1.0             1.0                      0.0        89.0   
3       1.0   1.0             0.0                      0.0        38.0   
4       1.0   0.0             1.0                      0.0        72.0   
..      ...   ...             ...                      ...         ...   
995     0.0   1.0             1.0                      1.0        88.0   
996     1.0   0.0             0.0                      0.0        56.0   
997     0.0   0.0             1.0                      1.0        59.0   
998     0.0   1.0             1.0                      1.0        68.0   
999     0.0   1.0             1.0                      0.0        77.0   

     reading_score  writing_score  reading_result  writing_result  math_result  
0             70.0           71.0             1.0             1.0          1.0  
1             87.0           84.0             1.0             1.0          1.0  
2             92.0           89.0             1.0             1.0          1.0  
3             57.0           44.0             0.0             0.0          0.0  
4             78.0           75.0             1.0             1.0          1.0  
..             ...            ...             ...             ...          ...  
995           97.0           91.0             1.0             1.0          1.0  
996           55.0           55.0             0.0             0.0          1.0  
997           69.0           62.0             1.0             1.0          0.0  
998           76.0           74.0             1.0             1.0          1.0  
999           83.0           82.0             1.0             1.0          1.0  

[1000 rows x 10 columns]

## 4) Use the two fairness metrics identified in 3.2 and compute the fairness metrics on the transformed dataset

In [ ]:
transformed.head()

gender  race  overall_result  test_preparation_course  math_score  \
0     0.0   1.0             1.0                      0.0        72.0   
1     0.0   0.0             1.0                      1.0        68.0   
2     0.0   1.0             1.0                      0.0        89.0   
3     1.0   1.0             0.0                      0.0        38.0   
4     1.0   0.0             1.0                      0.0        72.0   

   reading_score  writing_score  reading_result  writing_result  math_result  
0           70.0           71.0             1.0             1.0          1.0  
1           87.0           84.0             1.0             1.0          1.0  
2           92.0           89.0             1.0             1.0          1.0  
3           57.0           44.0             0.0             0.0          0.0  
4           78.0           75.0             1.0             1.0          1.0

In [ ]:
transformed['new_result'] = transformed.apply(lambda x: 1 if x['math_score'] >= 60 else 0, axis=1)

transformed.head()

gender  race  overall_result  test_preparation_course  math_score  \
0     0.0   1.0             1.0                      0.0        72.0   
1     0.0   0.0             1.0                      1.0        68.0   
2     0.0   1.0             1.0                      0.0        89.0   
3     1.0   1.0             0.0                      0.0        38.0   
4     1.0   0.0             1.0                      0.0        72.0   

   reading_score  writing_score  reading_result  writing_result  math_result  \
0           70.0           71.0             1.0             1.0          1.0   
1           87.0           84.0             1.0             1.0          1.0   
2           92.0           89.0             1.0             1.0          1.0   
3           57.0           44.0             0.0             0.0          0.0   
4           78.0           75.0             1.0             1.0          1.0   

   new_result  
0           1  
1           1  
2           1  
3           0  
4           1

#### Gender

In [ ]:
# for dependent variable of math_result
print(transformed.groupby(["gender", "new_result"]).agg({"new_result": "count"}))
print()
# Disparate Impact Ratio

favarouble_unpriv = 321 / (197 + 321)
favarouble_priv = 284 / (284 + 198)

_dir = favarouble_unpriv / favarouble_priv
print("DIR: ", _dir)

# Statistical Parity Difference

_spd = (favarouble_unpriv * 100) - (favarouble_priv * 100)
print("SPD: ", _spd)

                   new_result
gender new_result            
0.0    0                  197
       1                  321
1.0    0                  198
       1                  284

DIR:  1.0517292946870411
SPD:  3.047950143385819


In [ ]:
# the DIR value has improved! 

#### Gender - Overall_result


In [ ]:

transformed['over_all_new_result'] = transformed.apply(lambda x: 1 if x['math_score']+x['reading_score']+x['writing_score'] >= 180 else 0, axis=1)
# for dependent variable of overall_result
print(transformed.groupby(["gender", "over_all_new_result"]).agg({"over_all_new_result": "count"}))
print()
# Disparate Impact Ratio

favarouble_unpriv = 379 / (379 + 139)
favarouble_priv = 294 / (294 + 188)

_dir = favarouble_unpriv / favarouble_priv
print("DIR: ", _dir)

# Statistical Parity Difference

_spd = (favarouble_unpriv * 100) - (favarouble_priv * 100)
print("SPD: ", _spd)

                            over_all_new_result
gender over_all_new_result                     
0.0    0                                    139
       1                                    379
1.0    0                                    188
       1                                    294

DIR:  1.199524597483781
SPD:  12.170172543616523


# STEP 4


## 1) Randomly split your original dataset into training and testing datasets

Here, from the original dataset, we choose 'overall_result' as the outcome / dependent variable.

In [ ]:
features = ['gender','race','test_preparation_course','math_score','reading_score','writing_score','math_result','reading_result','writing_result']

X_original = data.loc[:, features]
X_original = X_original.convert_dtypes()
y_original = data.loc[:, ['overall_result']]
y_original = y_original.convert_dtypes()
X_original_train, X_original_test, y_original_train, y_original_test = train_test_split(X_original, y_original, random_state=randrange(100), train_size = 0.65, shuffle=True)

## 2) Randomly split your transformed dataset into training and testing datasets

In [ ]:
X_transformed = transformed.loc[:, features]
X_transformed = X_transformed.convert_dtypes()
y_transformed = transformed.loc[:, ['overall_result']]
y_transformed = y_transformed.convert_dtypes()
X_transformed_train, X_transformed_test, y_transformed_train, y_transformed_test = train_test_split(X_transformed, y_transformed, random_state=randrange(100), train_size = 0.65)

## 3) Train a classifier using the original training dataset from Step 4.1; select one of your dependent variables as the output label to train your classifier.

In [ ]:
logisticRegr = LogisticRegression(solver='liblinear')

logisticRegr.fit(X_original_train, y_original_train['overall_result'])

predictions_original = logisticRegr.predict(X_original_test)

score = logisticRegr.score(X_original_test, y_original_test)
print(f'The accuracy of the model on the original dataset is {score}')

The accuracy of the model on the original dataset is 0.9857142857142858


## 4) Train a classifier using the transformed training dataset from Step 4.2; select one of your dependent variables as the output label to train your classifier.

In [ ]:
logisticRegr = LogisticRegression(solver='liblinear')

logisticRegr.fit(X_transformed_train, y_transformed_train['overall_result'])

predictions_transformed = logisticRegr.predict(X_transformed_test)

score = logisticRegr.score(X_transformed_test, y_transformed_test)
print(f'The accuracy of the model on the transformed dataset is {score}')

The accuracy of the model on the transformed dataset is 0.9857142857142858


## 5) Select the privileged/unprivileged groups associated with one of your protected class variables (from Step 3.1); Use the two fairness metrics identified in Step 3.2 and and compute the fairness metrics for the classifier output associated with the original testing dataset and the transformed testing dataset

###### We select Gender as the protected class variables, and Male as the priviledged group and female as the unpriviledged group.

###### The two fairness Metrics used are Disparate Impact Ratio (DIR) and Statistical Parity Difference (SPD) 

In [ ]:
# First we compute the metrics for the classifier output associated with original testing dataset

# Step 1 Combine the predictions with the original testing dataset
X_original_test['overall_result'] = predictions_original


print(X_original_test.groupby(["gender", "overall_result"]).agg({"test_preparation_course": "count"}))
print()
# Disparate Impact Ratio

favarouble_unpriv = 160 / (160 + 32)
favarouble_priv = 104 / (104 + 54)

_dir = favarouble_unpriv / favarouble_priv
print("DIR: ", _dir)

# Statistical Parity Difference

_spd = (favarouble_unpriv * 100) - (favarouble_priv * 100)
print("SPD: ", _spd)


                       test_preparation_course
gender overall_result                         
0      0.0                                  28
       1.0                                 151
1      0.0                                  70
       1.0                                 101

DIR:  1.2660256410256412
SPD:  17.510548523206765


In [ ]:
# Second we compute the metrics for the classifier output associated with transformed testing dataset

# Step 1 Combine the predictions with the original testing dataset
X_transformed_test['overall_result'] = predictions_transformed


print(X_transformed_test.groupby(["gender", "overall_result"]).agg({"test_preparation_course": "count"}))
print()
# Disparate Impact Ratio

favarouble_unpriv = 154 / (154 + 30)
favarouble_priv = 90 / (90 + 76)

_dir = favarouble_unpriv / favarouble_priv
print("DIR: ", _dir)

# Statistical Parity Difference

_spd = (favarouble_unpriv * 100) - (favarouble_priv * 100)
print("SPD: ", _spd)


                       test_preparation_course
gender overall_result                         
0      0.0                                  26
       1.0                                 145
1      0.0                                  64
       1.0                                 115

DIR:  1.5437198067632851
SPD:  29.47878470403353


## 6) For each fairness metric, in table format, indicate if there were any differences in the outcomes for the privileged versus unprivileged group. Was there a positive change, negative change, or no change on that fairness metric after transforming the dataset (from Step 3.4)? Was there a positive change, negative change, or no change on that fairness metric after training the classifier - with respect to the original testing dataset and the transformed testing dataset?

#### Values for DIR



|             | DIR Value   | Change      |
| ----------- | ----------- | ----------- |
| Original Dataset      |  1.3138316294462473      |    NA    |
| Transformed Dataset   | 1.199524597483781      | Negative         |
| Original Testing Dataset(From Classifier)           | 1.2660256410256412            |   NA          |
|Transformed Testing Dataset(From Classifier)|1.5437198067632851| Positive

For the original and transformed dataset (before classification), the DIR value decreases. This shows that the bias in the dataset decreases after transformation.

For the testing datasets (after classifier), the DIR value actually increases. This shows that the bias in the system actually increases. This indicates that the given Classifier is not very helpful in removing bias.

#### Values for SPD

|             | SPD Value   | Change      |
| ----------- | ----------- | ----------- |
| Original Dataset      |  19.59819926624907     |    NA    |
| Transformed Dataset   | 12.170172543616523        | Negative         |
| Original Testing Dataset(From Classifier)           | 17.510548523206765            |   NA          |
|Transformed Testing Dataset(From Classifier)|29.47878470403353| Positive

For the original and transformed dataset (before classification), the SPD value decreases. This shows that the bias in the dataset decreases after transformation.

For the testing datasets (after classifier), the SPD value actually increases. This shows that the bias in the system actually increases. This indicates that the given Classifier is not very helpful in removing bias.

We obtain almost the same results with DIR.